### DISCUSSION TITLE
Enabling gPTP for Arduino Portenta H7 with the portenta hat carrier

### DISCUSSION THREAD

**Discussion by El-Bucheron** [zephyrproject-rtos/zephyr#80567] [Q&A]
Hello everyone, 

I am currently working on enabling the gPTP protocol for [Arduino Portenta H7](https://docs.zephyrproject.org/latest/boards/arduino/portenta_h7/doc/index.html) with the [Portenta Hat Carrier](https://docs.arduino.cc/hardware/portenta-hat-carrier/). I’m facing some issues and would appreciate some assistance. In summary, it seems that when gPTP is enabled, the board is unable to handle incoming Ethernet data, which prevents the gPTP state machine from initializing properly.

It’s important to mention that this configuration should support gPTP, as the Portenta's microcontroller is compatible, and the Portenta Hat Carrier provides an Ethernet port with hardware timestamping.

# Steps to Reproduce

1. **Modifying the Ethernet Devicetree**
Since the Portenta Hat Carrier isn’t directly supported by Zephyr, I made minor adjustments to the Ethernet devicetree for the Portenta to include the Ethernet port of the Hat Carrier. The small update allowed me to perform basic Ethernet operations like pinging and sending UDP packets:
```
&mac {
	status = "okay";
	pinctrl-0 = < &eth_ref_clk_pa1
		      &eth_crs_dv_pa7
		      &eth_rxd0_pc4
		      &eth_rxd1_pc5
		      &eth_tx_en_pg11
		      &eth_txd1_pg12
		      &eth_txd0_pg13 >;
	pinctrl-names = "default";
};

&mdio {
	status = "okay";
	pinctrl-0 = <&eth_mdio_pa2 &eth_mdc_pc1>;
	pinctrl-names = "default";

	ethernet-phy@0 {
		compatible = "ethernet-phy";
		reg = <0x00>;
		status = "okay";
	};
};
```

2. **Compiling and Flashing the gPTP Sample on the Portenta Board**
```
west build -b arduino_portenta_h7/stm32h747xx/m7 samples/net/gptp
```

3. **Starting the gPTP Daemon for linux**
```
sudo ./ptp4l -2 -f configs/gPTP.conf -i enp2s0 -m -q -l 6 -S
```

# Results
When running the code, the board gets stuck during the state machine initialization, continuously sending "Pdelay requests," as shown in the logs:
```
[00:46:07.912,000] <wrn> net_gptp: Reset Pdelay requests
[00:46:08.916,000] <wrn> net_gptp: Reset Pdelay requests
[00:46:09.921,000] <wrn> net_gptp: Reset Pdelay requests
[00:46:10.925,000] <wrn> net_gptp: Reset Pdelay requests
``` 
Wireshark gives us the following traffic over the Ethernet port between my computer ("Hewlett...") and the board ("STMicro"):
![Pasted image](https://github.com/user-attachments/assets/2812a774-b965-4ba3-b6da-2327ed033225)

# Analysis
Wireshark shows that while the Pdelay request is processed by my computer, and a response is sent, the board doesn’t seem to acknowledge it and sends instead a new request. A similar issue occurs when pinging the computer from the board. The board sends an ARP request, the computer responds, but the board sends another ARP request as if it hasn’t received the response. Pinging works fine when gPTP is disabled, indicating that the issue might be related to gPTP implementation.of this problem.

# Any suggestions ?
So far, I’ve tried tweaking the devicetree without success. I’m seeking assistance on how to resolve this issue—whether the solution lies in modifying gPTP drivers, the devicetree, or the source code. Any suggestions are welcome. Thank you in advance for your help!

**Comment by kartben** [comment]
Which version of Zephyr are you running?

**Comment by El-Bucheron** [comment]
Hello, I am currently using the version 3.7.99

**Comment by jukkar** [comment]
> Pinging works fine when gPTP is disabled, indicating that the issue might be related to gPTP implementation.of this problem.

Are you able to use other protocols than ICMP? Does UDP/TCP work with the board? (if you disable gPTP)
Just trying to understand if this a driver issue or is higher up in the stack. The gPTP works with various other boards so initial suspect is the driver.

**Comment by El-Bucheron** [comment]
Hello jukkar,

Yes, I am able to use UDP/TCP when gPTP is disabled, but not when it is enabled. 

When trying to send UDP packets to my computer from the board with gPTP enabled, the board sends first a ARP request to obtain the MAC address of my computer port, meaning it is actually capable of sending data. My computer then sends its MAC address to the board. But the board keeps sending the ARP request as it hasn't received the response from the computer. Moreover, the ARP cache is empty, which also suggests that the response form the computer hasn't been received. Inversely, when trying to send a UDP packet from my computer, it sends a ARP request to board which doesn't answer to it, also suggesting that the board can't read the data.

I therefore believe that enabling gPTP change something at a driver level which makes the board able to send data but enable the receive ones.

I hope that I have been precise enough.
